server/src/main/java/org/elasticsearch/index/mapper/ConcreteMappedFieldType.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.mapper;
+
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.DocValuesFieldExistsQuery;
+import org.apache.lucene.search.NormsFieldExistsQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.TermQuery;
+import org.elasticsearch.index.query.QueryShardContext;
+
+import java.util.Map;
+
+public abstract class ConcreteMappedFieldType extends MappedFieldType {
```

Comment:
Can you add some javadoc just to explain why we need this intermediate class?

Reply:
I am still hoping that we don't need this class, do you have thoughts? :) I will add javadocs if we decide to keep it.

Reply:
I'm easy either way, I think.  It might be that adding a base class test would also act as a reminder to implementers?  Something like:
- build an existsQuery and parse a document with the relevant field
- if its a NormsFieldExistsQuery then check the doc contains a field with norms enabled
- if its a DocValuesFieldExistsQuery then check the doc contains a doc values field
- if its a TermQuery then check the doc contains a `field_names` field
- otherwise we expect a MatchAllDocsQuery or MatchNoDocsQuery

Then mapper implementations that don't actually index anything will throw an error if they don't return a MatchAllDocsQuery or MatchNoDocsQuery

Reply:
I like the idea!

Reply:
I pushed an update that has a new test added to `MapperTestCase` for this. I think you'll have ideas around how to make it better, it still has some rough edges:

- binary field is special because its exists query don't work unless the field is stored or has doc_values, we end up querying field_names even when it's not there, maybe we should change the implementation to return MatchNoDocsQuery in that case.
- we are testing only the default settings for each field type. It would be nice to test also the different combinations around omitNorms and hasDocValues but not all field types allow to set them which would require more specific methods to be overridden etc.